### ISSUE TITLE
v2 Project: Motivation and path.

### ISSUE THREAD

**Issue by ddavness** [HypheX/trello-v1#6]
# v2 Roadmap

## The problem

Currently, the code is in spaghetti form, with clear inefficiencies regarding the use of `HttpService` and a dangerously soft line between fetching data from Trello and sending data to Trello.
Also, there are inconsistencies in argument checking that we need to fix.

In the end, the refactoring that would be required would be too much for a remake to be a valid option

## Fundamental principles of v2

- Git-like approach to the problem that sits in three basic operations:
- - Fetch data from Trello;
- - Make changes locally;
- - Push data to Trello.

- A caching system in order to support larger-scale games;

- Gracious error handling for errors that might be out of user control (404's, 429's, 500's);

- Proper error messages that are easy to understand (I recommend to see [this article](https://www.nngroup.com/articles/ten-usability-heuristics/)).

## Nice-to-haves:

- Compatibility with [Rojo](https://rojo.space);
- For [Roblox-TS](https://roblox-ts.github.io) users: An NPM package will be developed in parallel by me;
- We will begin also shipping v2 modules in the Roblox Website so that you don't have to snoop in the releases page.
- **ADDED 29TH JULY:** Module will be documented with [Documentation Reader](https://devforum.roblox.com/t/documentation-reader-a-plugin-for-scripters/128825) (by @ProbableAI) styling.

## Versioning:

Until a version is stable, we will name dev releases of v2 `2.0.0-dev.X`

**Comment by ddavness** [comment]
**ADDED 29TH JULY:** Module will be documented with [Documentation Reader](https://devforum.roblox.com/t/documentation-reader-a-plugin-for-scripters/128825) (by @ProbableAI) styling.

**Comment by LFS6502** [comment]
This was my first OOP project ever, so it is messy as heck.
v2 won't be rushed, like v1 was.

**Comment by ddavness** [comment]
First public preview: https://github.com/HypheX/roblox-trello/releases/tag/2.0.0-dev.6

Feel free to comment on my API design choices.

**Comment by ddavness** [comment]
# Current Status

Ending 2019, while there wasn't too much time to progress on the development, there are some stuff that are already done, in particular, the constructors for our entities:

- Client (renamed from Entity)
- Board
- List
- Label

(Card couldn't make it before the new decade, sorry 😢)

`Commit` was renamed to `Push`, and I also implemented a `Pull` method, etc.

More info will be given when the card constructor is done and everything is properly tested.

**Comment by enescaglayan** [comment]
finally they started to work on it again after a huge break(months)

**Comment by enescaglayan** [comment]
Nvm it was updated 1 months ago they started to get a huge break again

**Comment by enescaglayan** [comment]
I bet we will see trello api v2 in 2021

**Comment by ddavness** [comment]
Soon:tm:

I'm currently in exam season + my laptop is going to be sent for repair so I don't have much to work with at the moment.

**Comment by RonekoKitty** [comment]
no one cares
